### PR TITLE
fix: child span not acquiring transaction lock in some cases 

### PR DESCRIPTION
**Pull request by kruskall**
Spans from the same transaction share the dropped spans map. If they end concurrency a race condition could happen when updating the map.
Add a RW lock to prevent that.

Closes https://github.com/elastic/apm-agent-go/issues/1484

### CODE REVIEW

**Review by dmathieu — APPROVED** [review]

**Review by marclop — COMMENTED** [review]
Would you mind adding a test covering the bug?

**Comment by kruskall** [comment]
I've added a test and discovered more issues. It seems this is affecting more than just the dropped spans.

I've pushed some changes and added a test.

File: span.go [hunk 1]
```diff
@@ -397,6 +397,10 @@ func (s *Span) End() {
 
 		if s.Type != parentType || s.Subtype != parentSubtype {
 			s.dropWhen(true)
+			if s.tx != nil {
+				s.tx.mu.Lock()
+				defer s.tx.mu.Unlock()
```
Comment (marclop) [inline span.go:402]: Reading the godoc in `span.end()`, `s.tx.transactionData.mu` also needs to be held:

https://github.com/elastic/apm-agent-go/blob/326e364d6e52ea9601deb41a0004b633e1a03051/span.go#L466-L471

```suggestion
				defer s.tx.mu.Unlock()
				if !s.tx.ended() {
					s.tx.TransactionData.mu.Lock()
					defer s.tx.TransactionData.mu.Unlock()
				}
```

File: span_test.go [hunk 2]
```diff
@@ -1044,6 +1044,40 @@ func TestCompressSpanSameKindConcurrent(t *testing.T) {
...
+			})
+			child.Context.SetDestinationService(apm.DestinationServiceSpanContext{Resource: fmt.Sprintf("foo%d", i)})
+			child.End()
+			wg.Done()
+		}(i)
+	}
+
+	// Wait until all the spans have ended.
+	wg.Wait()
+
```
Comment (marclop) [inline span_test.go:1077]: The parent and transaction should also end:

```suggestion
	parent.End()
	tx.End()
```

File: span_test.go [hunk 3]
```diff
@@ -1044,6 +1044,40 @@ func TestCompressSpanSameKindConcurrent(t *testing.T) {
 	assert.Equal(t, 101, spanCount)
 }
 
+func TestDroppedSpanParentConcurrent(t *testing.T) {
+	// This test verifies there aren't any deadlocks on calling
+	// span.End(), Parent.End() and tx.End().
+	tracer := apmtest.NewRecordingTracer()
```
Comment (marclop) [inline span_test.go:1050]:
```suggestion
	tracer := apmtest.NewRecordingTracer()
	defer tracer.Close()
```

**Review by marclop — COMMENTED** [review]
Thanks, this looks better!

**Review by simitt — COMMENTED** [review]
Once this is ready to merge, please also add a changelog entry

**Comment by kruskall** [comment]
@simitt could you clarify the above question ?

**Comment by axw** [comment]
@kruskall in the past we have added entries to the Unreleased section. We've not been consistent about it though, and have sometimes just added the entries retrospectively at release time. Let's merge this and create a patch release - either adding the entry now, or creating the section at release time.

**Comment by simitt** [comment]
Apologies @kruskall , I missed your message.

**Review by marclop — APPROVED** [review]

**Pull request merged**